TestFoundation/TestOperationQueue.swift

```diff
@@ -168,34 +168,42 @@ class TestOperationQueue : XCTestCase {
 class AsyncOperation: Operation {
 
     private let queue = DispatchQueue(label: "async.operation.queue")
-    private let lock = NSLock()
+    private let lock = NSRecursiveLock()
```

Comment:
Right, it's a test and therefore I'd be happy with a recursive lock, generally speaking they lead to dead-lock prone code (as establishing a lock order isn't as easy anymore).

A `DispatchQueue` is very cheap and in ObjC at least a synchronous dispatch on a dispatch queue is basically the same speed as taking a mutex. They should be both > 5 ns and < 50 ns for a non-contended case. Said that, in Swift, there's currently a [performance issue with synchronous dispatches](https://bugs.swift.org/browse/SR-4937). This  shouldn't matter here as an Asynchronous Operation is a class anyway meaning we'll get memory allocations and reference counting. In my benchmarks it should be 20 ns (nano seconds) per synchronous dispatch but in Swift it's 140 ns. I'd be really surprised if you'd see this difference when using Operation/OperationQueue.

The advantage of a dispatch queue is that the API is easier to use (can't forget to unlock) and it's the preferred way of synchronising.